src/transformers/CommonUtils.cc

```diff
@@ -108,3 +108,127 @@ InvalidityWitness::Derivation replaceSummarizingStep(
     InvalidityWitness::Derivation newDerivation(std::move(newSteps));
     return newDerivation;
 }
+
+
+
+InvalidityWitness::Derivation expandStepWithHyperEdge(
+    InvalidityWitness::Derivation const & derivation,
+    std::size_t stepIndex,
+    ContractionInfo const & contractionInfo,
+    NonlinearCanonicalPredicateRepresentation const & predicateRepresentation,
+    Logic & logic
+    ) {
+    assert(stepIndex < derivation.size());
+    // Replace this step with the sequence of steps corresponding to the summarized chain
+    std::vector<DerivationStep> newSteps;
+    newSteps.reserve(stepIndex);
+    // 1. Copy all the steps before the one to be replaced
+    std::copy(derivation.begin(), derivation.begin() + stepIndex, std::back_inserter(newSteps));
+    std::size_t firstShiftedIndex = newSteps.size();
+    // 2. Replace the summarized step
+    DerivationStep const & summarizedStep = derivation[stepIndex];
+    // 2a. Compute instance for the contracted node
+    // 2aa. Compute path constraint
+    auto const & incoming = contractionInfo.second.first;
+    auto const & outgoing = contractionInfo.second.second;
+    auto contractedNode = incoming.to;
+    assert(std::count(outgoing.from.begin(), outgoing.from.end(), contractedNode) == 1); // It has to be there, and currently not more than once
+    assert(incoming.from.size() == 1);
+    bool sourceIsEntry = incoming.from.at(0) == logic.getSym_true();
+    PTRef constraint = logic.mkAnd(incoming.fla.fla, outgoing.fla.fla);
+    auto unifiedConstraint = [&](PTRef constraint) -> PTRef {
+        vec<PTRef> unifyingEqualities;
+        PTRef sourceTerm = predicateRepresentation.getSourceTermFor(contractedNode, 0);
+        PTRef targetTerm = predicateRepresentation.getTargetTermFor(contractedNode);
+        TermUtils::substitutions_map substitutionsMap;
+        TermUtils(logic).mapFromPredicate(sourceTerm, targetTerm, substitutionsMap);
+        return TermUtils(logic).varSubstitute(constraint, substitutionsMap);
+    }(constraint);
+    // 2ab Compute constraints for the other nodes
+    TermUtils::substitutions_map subst;
+    TermUtils utils(logic);
+    auto const & replacingEdge = contractionInfo.first;
+    assert(summarizedStep.premises.size() == replacingEdge.from.size());
+    assert(logic.getSymRef(summarizedStep.derivedFact) == replacingEdge.to);
+
+    const PTRef targetPredicate = predicateRepresentation.getTargetTermFor(replacingEdge.to);
+    utils.mapFromPredicate(targetPredicate, summarizedStep.derivedFact, subst);
+    {
+        // TODO: Better way to compute instances correctly?
+        std::unordered_map<SymRef, std::size_t, SymRefHash> instanceCounter;
```

Comment:
What is the point of instanceCounter exactly? It is only used as a parameter at line 164

Reply:
We need to correctly count the number of occurences of each vertex in the sources. If the same vertex occurs multiple times in the sources of the hyperedge, its variables must be distinguished for each occurence.